### PR TITLE
Theano batch normalization in 5D and with new gpu backend

### PR DESCRIPTION
**Pull request by gvtulder**
The first patch enables Theano's cuDNN function for batch normalization for 5D tensors, if this is supported by the Theano version. (See also #3666, where cuDNN was disabled for 5D because Theano did not support it at the time.)

The second patch uses the batch normalization functions from Theano's new gpuarray backend if device=cuda. This seems to run faster than the old theano.sandbox.cuda backend.

### CODE REVIEW

**Comment by nouiz** [comment]
The second commit have a snack mistake. Line 409 and 413 are reversed.

Le 23 oct. 2016 18:04, "Gijs van Tulder" <notifications@github.com> a
écrit :

**Comment by gvtulder** [comment]
@nouiz: Could you explain that a bit further, please? I might be missing something. I was under the impression that `device=gpu*` would use the old `theano.sandbox.cuda` backend, whereas `device=cuda*` uses the `theano.gpuarray` backend. That's exactly what lines 409 and 413 do, I think. If it's the opposite way around, lines and 456 and 460 probably need to be swapped too?

**Comment by nouiz** [comment]
The code seem right. I don't know I saw a problem last time. So all seem good from Theano side. Not sure about the integration in keras.

**Comment by gvtulder** [comment]
Rebased to the current master to fix a conflict with #3968.

**Comment by fchollet** [comment]
The code looks good to me. Don't have the means to test it, though (and Travis isn't testing it either - CPU only).

**Comment by gvtulder** [comment]
@nouiz: What would you think of adding abstract batch normalization ops to Theano? That is, two abstract ops that represent batch_normalization_train and batch_normalization_test, following the cuDNN semantics. Those ops can then be optimized into either one of the two gpu/cuDNN ops, or replaced with an implemetation using normal Theano ops for when cuDNN is not available.

Such functions would move much of this slightly awkward logic to Theano. For non-Keras purposes, this would also make it a lot easier to use these cuDNN BN functions, because you no longer have to write your code for a specific cpu/gpu backend.

The current `theano.tensor.nnet.bn.batch_normalization` is a bit different from the cuDNN functions, so I'd propose two new functions, for example, `theano.tensor.nnet.bn.batch_normalization_train` and `theano.tensor.nnet.bn.batch_normalization_test`, that follow exactly the cuDNN form.

(I can move this to a Theano issue, if you want.)

**Comment by nouiz** [comment]
This was moved to Theano and there is a PR that started it:

https://github.com/Theano/Theano/pull/5190

I think having the abstraction in Theano is good. But there is interface
questions, so it won't be like just moving the code from this PR to Theano.

**Comment by fchollet** [comment]
Cool. Will put this PR on hold until we can rely on the Theano-side fix.

**Comment by fchollet** [comment]
@nouiz do you know if this has been fixed in Theano at this point?

**Comment by gvtulder** [comment]
I have a new branch that uses the new Theano interface (which isn't included yet, but is working for me so far). I've just opened a new pull request #4595 for that work-in-progress. I'll close this PR because it's better to just upgrade to the new interface.
